### ISSUE TITLE
Move to an sdk package

### ISSUE THREAD

**Issue by lil5** [balzack/databag#95]
Instead of reimplementing common chat functionality on each client, create an sdk that is plugged into mobile and web apps. 

This could handle; 
- Improve bot support 🤖 
- Move to redaxios 

The sdk should not contain any react. Instead async and callbacks should be used

```TS
const sdk = new DatabagSdk()

sdk.onLogin(()=>{})

const room = sdk.selectRoom()
const g = () => appview(room.renderConversation())
room.onUpdate(g)
```

This is just an example of functionality not structure. 

With an sdk the react could move to a more passive role of displaying data when the sdk says so.

**Comment by balzack** [comment]
Yes, I agree. Along those lines, I can see it allowing for other clients to participate in the network while maintaining interoperability. Just because clients speak the same protocol or api, doesn't mean they work well together.

It would be nice if there was just one SDK, but mobile generally requires storage for offline viewing. Supporting both modes within a single SDK may be messy.

There are a few new features, maybe bundled as separate apps, that I would like to include in an SDK. Perhaps it's over designing, but it goes to my hopes that the databag backend can be more general than a message service.

- contact list with selective access to attributes (ie. phone number for friends, email for work colleagues)
- photo albums with selective access (ie. family can see family photos, work colleagues can see conference photos)

Thank you for the suggestion!

**Comment by lil5** [comment]
> but mobile generally requires storage for offline viewing. Supporting both modes within a single SDK may be messy.

That would require an optional `addOfflineSqliteStorage(db)`

Interoperability with other sqlite libraries isn't necessary as a non react native library exists too. https://github.com/asnando/sqlite-storage

**Comment by balzack** [comment]
Thank you for this reference. I have a rough idea what the API should look like now, but am not yet sure how much the refactor effort will be.

I will likely start the SDK work in a month or two after I complete several other commitments. Would you be willing to review the SDK API after I finish a draft?

**Comment by lil5** [comment]
Definitely 👍 

I'd suggest cut-pasting from frontend to sdk then altering from there, that way you never have a commit where everything is broken.

**Comment by balzack** [comment]
I need to move the typescript changes into a branch until I can fix it.

Deploying the latest gives me: 'Uncaught ReferenceError: require is not defined'

**Comment by lil5** [comment]
I'll see if I can work on the docker development env, you seem to be using production builds instead of the local code. 

I haven't yet got it working properly (on a Mac machine)

**Comment by balzack** [comment]
It might not be worth this effort. When I start the SDK refactor work, I will make sure to use typescript.

**Comment by balzack** [comment]
@lil5 - I pushed the refactored code to app/client. I'll start a branch to find a methodical way to migrate the code base over. In the meantime, please let me know if you have any stability or UI concerns. I have been testing for several weeks now and I believe it's production ready. Once this refactor is wrapped up, I can get back to adding features, at the top of the list deleting old messages, which I believe you mentioned.

the go code has a (cors) commit last week allowing a dev server to exist on another machine.
to build and launch the server on localhost:3000; from app/client/web:
> yarn install & yarn dev

**Comment by balzack** [comment]
The server name is defaulted to the dev server's name. Clicking on it allows you to select the proper server name.
